packages/grafana-data/src/utils/dataLinks.ts

```diff
@@ -12,3 +26,96 @@ export const DataLinkBuiltInVars = {
   // name of the calculation represented by the value
   valueCalc: '__value.calc',
 };
+
+type Options = {
+  onClickFn?: (options: { datasourceUid: string; query: any }) => void;
+  replaceVariables: InterpolateFunction;
+  getDataSourceSettingsByUid: (uid: string) => DataSourceInstanceSettings | undefined;
+};
+
+export function mapInternalLinkToExplore(
+  link: DataLink,
+  scopedVars: ScopedVars,
+  range: TimeRange,
+  field: Field,
+  options: Options
+): LinkModel<Field> {
+  if (!link.internal) {
+    throw new Error('Trying to map external link as internal');
+  }
+  const { onClickFn, replaceVariables, getDataSourceSettingsByUid } = options;
+
+  const interpolatedQuery = interpolateQuery(link, scopedVars, replaceVariables);
+  return {
+    title: link.title
+      ? replaceVariables(link.title || '', scopedVars)
+      : getDataSourceSettingsByUid(link.internal.datasourceUid)?.name || 'Unknown datasource',
+
+    // In this case this is meant to be internal link (opens split view by default) the href will also points
+    // to explore but this way you can open it in new tab.
+    href: generateInternalHref(
+      getDataSourceSettingsByUid(link.internal.datasourceUid)?.name || 'unknown',
+      interpolatedQuery,
+      range
+    ),
+    onClick: onClickFn
+      ? () => {
+          onClickFn?.({
+            datasourceUid: link.internal!.datasourceUid,
+            query: interpolatedQuery,
+          });
+        }
+      : undefined,
+    target: '_self',
+    origin: field,
+  };
+}
+
+/**
+ * Generates href for internal derived field link.
+ */
+function generateInternalHref<T extends DataQuery = any>(datasourceName: string, query: T, range: TimeRange): string {
+  return locationUtil.assureBaseUrl(
+    `/explore?left=${serializeStateToUrlParam({
+      range: range.raw,
+      datasource: datasourceName,
+      queries: [query],
+      // This should get overwritten if datasource does not support that mode and we do not know what mode is
+      // preferred anyway.
+      mode: ExploreMode.Metrics,
+      ui: {
+        showingGraph: true,
+        showingTable: true,
+        showingLogs: true,
+      },
+    })}`
+  );
+}
+
+function interpolateQuery<T extends DataQuery = any>(
+  link: DataLink,
+  scopedVars: ScopedVars,
+  replaceVariables: InterpolateFunction
+): T {
+  let stringifiedQuery = '';
+  try {
+    stringifiedQuery = JSON.stringify(link.internal?.query || '');
```

Comment:
Not sure if this is the best way to interpolate variables inside random query object. One caveat is that this will replace it also in keys not only values.